File: LISTAS/LISTA DOBLEMENTE ENLAZADA/LISTADOBLEMENTEENLAZADA.java
public class ListaDoblementeEnlazada {
    class NodoDoble {
        int dato;
        NodoDoble siguiente;
        NodoDoble anterior;
        
        NodoDoble(int dato) {
            this.dato = dato;
            this.siguiente = null;
            this.anterior = null;
        }
    }
    
    private NodoDoble cabeza;
    private NodoDoble cola;
    
    public ListaDoblementeEnlazada() {
        this.cabeza = null;
        this.cola = null;
    }
    
    public void insertarInicio(int dato) {
        NodoDoble nuevo = new NodoDoble(dato);
        if (cabeza == null) {
            cabeza = cola = nuevo;
        } else {
            nuevo.siguiente = cabeza;
            cabeza.anterior = nuevo;
            cabeza = nuevo;
        }
    }
    
    public void insertarFinal(int dato) {
        NodoDoble nuevo = new NodoDoble(dato);
        if (cabeza == null) {
            cabeza = cola = nuevo;
        } else {
            nuevo.anterior = cola;
            cola.siguiente = nuevo;
            cola = nuevo;
        }
    }
    
    public void eliminar(int dato) {
        NodoDoble actual = cabeza;
        while (actual != null) {
            if (actual.dato == dato) {
                if (actual.anterior != null) {
                    actual.anterior.siguiente = actual.siguiente;
                } else {
                    cabeza = actual.siguiente;
                }
                
                if (actual.siguiente != null) {
                    actual.siguiente.anterior = actual.anterior;
                } else {
                    cola = actual.anterior;
                }
                return;
            }
            actual = actual.siguiente;
        }
    }
    
    public void mostrarAdelante() {
        NodoDoble actual = cabeza;
        while (actual != null) {
            System.out.print(actual.dato + " ");
            actual = actual.siguiente;
        }
        System.out.println();
    }
    
    public void mostrarAtras() {
        NodoDoble actual = cola;
        while (actual != null) {
            System.out.print(actual.dato + " ");
            actual = actual.anterior;
        }
        System.out.println();
    }
}

File: LISTAS/LISTA DOBLEMENTE ENLAZADA/LISTADOBLEMENTEENLAZADA.js
class NodoDoble {
    constructor(dato) {
        this.dato = dato;
        this.siguiente = null;
        this.anterior = null;
    }
}

class ListaDoblementeEnlazada {
    constructor() {
        this.cabeza = null;
        this.cola = null;
    }
    
    insertarInicio(dato) {
        const nuevo = new NodoDoble(dato);
        if (!this.cabeza) {
            this.cabeza = this.cola = nuevo;
        } else {
            nuevo.siguiente = this.cabeza;
            this.cabeza.anterior = nuevo;
            this.cabeza = nuevo;
        }
    }
    
    insertarFinal(dato) {
        const nuevo = new NodoDoble(dato);
        if (!this.cabeza) {
            this.cabeza = this.cola = nuevo;
        } else {
            nuevo.anterior = this.cola;
            this.cola.siguiente = nuevo;
            this.cola = nuevo;
        }
    }
    
    eliminar(dato) {
        let actual = this.cabeza;
        while (actual) {
            if (actual.dato === dato) {
                if (actual.anterior) {
                    actual.anterior.siguiente = actual.siguiente;
                } else {
                    this.cabeza = actual.siguiente;
                }
                
                if (actual.siguiente) {
                    actual.siguiente.anterior = actual.anterior;
                } else {
                    this.cola = actual.anterior;
                }
                return;
            }
            actual = actual.siguiente;
        }
    }
    
    mostrarAdelante() {
        const elementos = [];
        let actual = this.cabeza;
        while (actual) {
            elementos.push(actual.dato);
            actual = actual.siguiente;
        }
        return elementos;
    }
    
    mostrarAtras() {
        const elementos = [];
        let actual = this.cola;
        while (actual) {
            elementos.push(actual.dato);
            actual = actual.anterior;
        }
        return elementos;
    }
}